[AutoDiff upstream] Add `@derivative(of:)` attribute.

dan-zheng opened the pull request:
The `@derivative(of:)` attribute registers a function as a derivative of another function.
This patch adds the `@derivative(of:)` attribute definition, syntax, parsing, and printing.

Resolves TF-826.

Todos:
- Type-checking (TF-829).
- Serialization (TF-837).

dan-zheng commented:
cc @rxwei @marcrasi @bgogul

lattner commented on include/swift/AST/Attr.def:
```diff
@@ -508,6 +508,11 @@ DECL_ATTR(differentiable, Differentiable,
   ABIStableToAdd | ABIBreakingToRemove | APIStableToAdd | APIBreakingToRemove,
   91)
 
+DECL_ATTR(differentiating, Differentiating,
```
Bikeshed language design issue - feel free to punt to the formal review process, but this attribute is textually very similar to @differentiable.  Did you consider something like `@derivativeOf` as the name of this attribute?

rxwei replied:
I'm very open to this idea! Personally, I like `@derivative(of: ...)` and `@transpose(of: ...)` better than `@differentiating(...)` and `@transposing(...)`.

lattner replied:
+1 for `@derivative(of: ...)`

dan-zheng replied:
`@derivative(of: ...)` does clearly evoke derivative registration! I'll rename to `@differentiating(...)` to `@derivative(of: ...)`, and deprecate `@differentiating(...)` on `tensorflow` branch with a warning, if there are no objections.

saeta replied:
+1

dan-zheng replied:
Renamed `@differentiating` to `@derivative(of:)` in https://github.com/apple/swift/pull/28321/commits/4b9f025351afef770a431ae276581d5cf473e736.

lattner reviewed (approved):
LGTM with a few changes, thanks!

rxwei commented on lib/AST/Attr.cpp:
```diff
@@ -864,6 +864,20 @@ bool DeclAttribute::printImpl(ASTPrinter &Printer, const PrintOptions &Options,
+    auto *attr = cast<DifferentiatingAttr>(this);
+    auto *derivative = dyn_cast_or_null<AbstractFunctionDecl>(D);
+    Printer << attr->getOriginal().Name;
+    auto diffParamsString = getDifferentiationParametersClauseString(
```
Completely agreed. This should definitely be `cast<>`. `getDifferentiationParametersClauseString`'s semantics are unclear regarding whether the first argument can be null -- it appears that it is dereferencing the pointer without any check at one point.

dan-zheng replied:
Thanks for the suggestions! I addressed feedback in https://github.com/apple/swift/pull/28377:
* Clarified doc comments and parameter non-nullability for `getDifferentiationParametersClauseString`.
* Changed existing occurrence of `dyn_cast_or_null<AbstractFunctionDecl>(D)` to `cast<AbstractFunctionDecl>(D)`.

dan-zheng commented on lib/Parse/ParseDecl.cpp:
```diff
@@ -1059,6 +1059,80 @@ bool Parser::parseDifferentiableAttributeArguments(
+    }
+    // Check that token after comma is 'linear' or 'wrt:'.
+    if (!Tok.is(tok::identifier) ||
+        !(Tok.getText() == "linear" || Tok.getText() == "wrt")) {
```
> This logic is very difficult to read with all the negation happening, I'd recommend inverting this to be "if (positive test) return false; emit error"

Refactored existing code to use simpler positive conditions in https://github.com/apple/swift/pull/28377. I'll update this PR too.

dan-zheng committed:
[AutoDiff upstream] Add `@derivative(of:)` attribute.

The `@derivative(of:)` attribute registers a function as a derivative of another
function. This patch adds the `@derivative(of:)` attribute definition, syntax,
parsing, and printing.

Resolves TF-826.

Todos:
- Type-checking (TF-829).
- Serialization (TF-837).

dan-zheng commented on lib/Parse/ParseDecl.cpp:
```diff
@@ -1059,6 +1059,80 @@ bool Parser::parseDifferentiableAttributeArguments(
   return false;
 }
 
+ParserResult<DifferentiatingAttr>
```
Done.

Also added production rules for `Parser::parseDifferentiationParametersClause` and `Parser::parseDifferentiableAttribute`.

dan-zheng commented:
I addressed all comments and rebased onto master.

There are no major changes: the biggest syntax change is renaming `@differentiating` to `@derivative(of:)`, which has already been done on `tensorflow` branch: https://github.com/apple/swift/pull/28481.

dan-zheng commented:
@swift-ci Please test
https://github.com/apple/swift-syntax/pull/188

dan-zheng commented:
@swift-ci Please test
apple/swift-syntax#188

dan-zheng commented:
apple/swift-syntax#188
@swift-ci Please test

dan-zheng commented:
Merging since there are no major changes since the last approval and CI has passed.
